Enable KernelShap support for string type input

YotamElor opened the pull request:
SKLearn pipelines use encoders to support string input e.g. OneHotEncoder and OrdinalEncoder. In some cases it is desired to calculate the shap values in the original input space i.e. before encoding it into numeric values. While there is no theoretical limitation, currently KernelExplainer does not support string type input due to the limitations of a single utility function. The utility function varying_groups() relies on numpy functions that assume numeric input. After fixing it to support string type input, KernelExplainer fully supports string values.

YotamElor committed:
Enable KernelShap support for string type input

YotamElor committed:
Enable KernelShap support for string type input

niloofardadras commented:
@YotamElor  hi, has this been implemented yet?
I am looking to use string features for kernel explainer and I am not able to do anything except label encoding but shap results do not make sense for them. please let me know if there is any solution for it.

YotamElor commented:
Yes, it is implemented, feel free to use this code

niloofardadras commented:
@YotamElor i have tried make the change in package but it gave following error. i use tensorflow and my pipeline 1.uses dataframe with both string and floats 2. passes dataframe to create input tensor of string and floats 3. passes the input for prediction. i get following error when i make the changes you have in kernel.py.

"tensorflow.python.framework.errors_impl.InternalError: Unable to get element as bytes"

lrjball committed:
Fixes error with XGBoost 1.1

In the new XGBoost (1.1), `.save_raw()` output is slightly different, which means that `TreeExplainer` doesn't work. See issue #1215. 
The build is currently failing for this reason, with all XGBoost tests failing.

The fix here is to just remove the prefix 'binf' from the model buffer, which as been added in the new version (either intentionally or accidentally - I can't find a PR in XGBoost which mentions this change.)

Fixes #1215

slundberg committed:
Merge pull request #1220 from lrjball/patch-1

Fixes error with XGBoost 1.1

slundberg committed:
Add sum test

slundberg committed:
Add sum test

slundberg commented:
Thanks @YotamElor ! I added a quick sum test to the unit test. Assuming that checks out I'll go ahead and merge this :)

niloofardadras commented:
@slundberg   Hi scott,
is there any plan to have support for string type input with deep explainer maybe? as i have tried seems tonot be supported with deep explainer also.

Best regards

YotamElor committed:
Fixed the sum test. Added an assertion verifying that the shap value for a constant string feature is zero

YotamElor committed:
Fixed the sum test. Added an assertion verifying that the shap value for a constant string feature is zero

YotamElor committed:
Rebasing in order to pass checks. Merge branch 'master' of github.com:YotamElor/shap

rchatwin1 commented:
@YotamElor: Your not_equal function does not handle missing data. If you call not_equal(i, j) with i = nan, then isinstance(i, str) will resolve to False. If at the same time j is a string, then the call to np.isclose will fail. This is a common use case when using pandas dataframes where all null data is set to nan regardless of whether the column data type is 'object' or some numeric type. 

Also, the behavior with missing data is different when comparing strings than when comparing numerics: with numerics, not_equal returns 1 when one entry is missing but the other is not, whereas with strings not_equal returns 0 if either entry is None.

YotamElor committed:
Fix the behavior for np.nan and None values. Added tests to verify consistent behavior

YotamElor commented:
@rchatwin1 , thank you for noting that. I fixed that behavior and added tests to verify consistent behavior for string and numeric values
